Index: openocd/configure.in
===================================================================
--- openocd/configure.in	(revision 1279)
+++ openocd/configure.in	(working copy)
@@ -247,7 +247,11 @@
   AS_HELP_STRING([--enable-rlink], [Enable building support for the Raisonance RLink JTAG Programmer]),
   [build_rlink=$enableval], [build_rlink=no])
 
+AC_ARG_ENABLE(idbg,
+  AS_HELP_STRING([--enable-idbg], [Enable building support for the IDBG JTAG Programmer]),
+  [build_idbg=$enableval], [build_idbg=no])
 
+
 case $host in 
   *-cygwin*) 
     is_win32=yes
@@ -418,6 +422,13 @@
   AC_DEFINE(BUILD_RLINK, 0, [0 if you don't want the RLink JTAG driver.])
 fi
 
+if test $build_idbg = yes; then
+  build_bitbang=yes
+  AC_DEFINE(BUILD_IDBG, 1, [1 if you want the IDBG JTAG driver.])
+else
+  AC_DEFINE(BUILD_IDBG, 0, [0 if you don't want the IDBG JTAG driver.])
+fi
+
 #-- Deal with MingW/Cygwin FTD2XX issues
 
 if test $is_win32 = yes; then
@@ -632,6 +643,7 @@
 AM_CONDITIONAL(OOCD_TRACE, test $build_oocd_trace = yes)
 AM_CONDITIONAL(JLINK, test $build_jlink = yes)
 AM_CONDITIONAL(RLINK, test $build_rlink = yes)
+AM_CONDITIONAL(IDBG, test $build_idbg = yes)
 AM_CONDITIONAL(IS_CYGWIN, test $is_cygwin = yes)
 AM_CONDITIONAL(IS_MINGW, test $is_mingw = yes)
 AM_CONDITIONAL(IS_WIN32, test $is_win32 = yes)
Index: openocd/src/jtag/idbg.c
===================================================================
--- openocd/src/jtag/idbg.c	(revision 0)
+++ openocd/src/jtag/idbg.c	(revision 0)
@@ -0,0 +1,334 @@
+/*
+ * idbg.c - IDBG JTAG interface
+ *
+ * Written 2008 by Werner Almesberger
+ * Copyright 2008 Werner Almesberger
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; either version 2 of the License, or
+ * (at your option) any later version.
+ */
+
+/*
+ * This is the most primitive way to use IDBG. We basically implement a
+ * USB-attached wiggler. Performance cannot be anything but horrible, but this
+ * is a great way to get started.
+ */
+
+
+#include <stdint.h>
+#include <usb.h>
+
+#include "jtag.h"
+
+
+#define TO_DEV		0x40
+#define	FROM_DEV	0xc0
+
+#define IDBG_TO_DEV(req)        (0x40 | (req) << 8)
+#define IDBG_FROM_DEV(req)      (0xc0 | (req) << 8)
+
+/*
+ * @@@ The real limit for MAX_SCAN_BYTES should be 64, but we may hit some EP
+ * size issues when going to this limit. See the corresponding DFU upload
+ * problem. So fix DFU upload first.
+ */
+
+#define MAX_SCAN_BYTES	63
+
+
+enum idbg_requests {
+	IDBG_ID			= 0x00,
+	IDBG_RESET,
+	IDBG_JTAG_ATTACH	= 0x10,
+	IDBG_JTAG_DETACH,
+	IDBG_JTAG_GPIO_SET,
+	IDBG_JTAG_GPIO_GET,
+	IDBG_JTAG_SCAN,
+	IDBG_JTAG_FETCH,
+	IDBG_JTAG_MOVE,
+};
+
+enum jtag_gpios {
+	JTAG_nTRST,	/* 0x01 */
+	JTAG_TMS,	/* 0x02 */
+	JTAG_TDO,	/* 0x04 */
+	JTAG_TDI,	/* 0x08 */
+	JTAG_TCK,	/* 0x10 */
+	JTAG_nSRST,	/* 0x20 */
+};
+
+
+static usb_dev_handle *usb;
+
+
+static usb_dev_handle *open_usb(void)
+{
+	const struct usb_bus *bus;
+	struct usb_device *dev;
+
+	usb_init();
+	usb_find_busses();
+	usb_find_devices();
+
+	for (bus = usb_get_busses(); bus; bus = bus->next)
+		for (dev = bus->devices; dev; dev = dev->next) {
+			if (dev->descriptor.idVendor != 0x1234)
+				continue;
+			if (dev->descriptor.idProduct != 0x0002)
+				continue;
+			return usb_open(dev);
+		}
+	return NULL;
+}
+
+
+static void idbg_write(int tck, int tms, int tdi)
+{
+	int res;
+
+	LOG_DEBUG("set tck %d tms %d tdi %d", tck, tms, tdi);
+	res = usb_control_msg(usb, TO_DEV, IDBG_JTAG_GPIO_SET,
+	    !!tck << JTAG_TCK | !!tms << JTAG_TMS | !!tdi << JTAG_TDI,
+	    1 << JTAG_TCK | 1 << JTAG_TMS | 1 << JTAG_TDI,
+	    NULL, 0, 1000);
+	if (res < 0) {
+		LOG_ERROR("IDBG_JTAG_GPIO_SET failed: %i", res);
+		exit(1);
+	}
+}
+
+static void idbg_reset(int trst, int srst)
+{
+	int res;
+
+	LOG_DEBUG("set trst %d srst %d", trst, srst);
+	res = usb_control_msg(usb, TO_DEV, IDBG_JTAG_GPIO_SET,
+	    !trst << JTAG_nTRST | !srst << JTAG_nSRST,
+	    1 << JTAG_nTRST | 1 << JTAG_nSRST,
+	    NULL, 0, 1000);
+	if (res < 0) {
+		LOG_ERROR("IDBG_JTAG_GPIO_SET failed: %i", res);
+		exit(1);
+	}
+}
+
+
+static void idbg_move(uint8_t scan)
+{
+	int res;
+
+	LOG_DEBUG("move: 0x%2x", scan);
+	res = usb_control_msg(usb, TO_DEV, IDBG_JTAG_MOVE,
+	    7, 0, &scan, 1, 1000);
+	if (res < 0) {
+		LOG_ERROR("IDBG_JTAG_MOVE failed: %i", res);
+		exit(1);
+	}
+}
+
+
+static void idbg_state_move(enum tap_state state)
+{
+	uint8_t scan = TAP_MOVE(cur_state, state);
+
+	LOG_DEBUG("idbg_state_move %d -> %d", cur_state, state);
+	if (cur_state == state)
+		return;
+	idbg_move(scan);
+	cur_state = state;
+}
+
+
+static void idbg_path_move(pathmove_command_t *cmd)
+{
+	uint8_t scan = 0;
+	enum tap_state next;
+	int i, tms;
+
+	for (i = 0; i != cmd->num_states; i++) {
+		next = cmd->path[i];
+		tms = tap_transitions[cur_state].low == next;
+		scan |= tms << i;
+		cur_state = next;
+	}
+	idbg_move(scan);
+}
+
+
+void idbg_runtest(int num_cycles)
+{
+	int i;
+
+	LOG_DEBUG("runtest %d", num_cycles);
+	idbg_state_move(TAP_IDLE);
+	for (i = 0; i != num_cycles; i++) {
+		idbg_write(0, 0, 0);
+		idbg_write(1, 0, 0);
+	}
+	idbg_write(0, 0, 0);
+	idbg_state_move(end_state);
+}
+
+
+static void idbg_scan(int ir_scan, enum scan_type type, u8 *buffer,
+    int scan_size)
+{
+	int bit = 0, end, bytes, res;
+
+	LOG_DEBUG("scan %d in %d", scan_size, type != SCAN_OUT);
+	idbg_state_move(ir_scan ? TAP_IRSHIFT : TAP_DRSHIFT);
+	while (bit != scan_size) {
+		end = bit+MAX_SCAN_BYTES*8;
+		if (end > scan_size)
+			end = scan_size;
+		bytes = (end-bit+7) >> 3;
+		res = usb_control_msg(usb, TO_DEV, IDBG_JTAG_SCAN,
+		    end-bit, end == scan_size, buffer, bytes, 1000);
+		if (res < 0) {
+			LOG_ERROR("IDBG_JTAG_SCAN failed: %i", res);
+			exit(1);
+		}
+		if (type != SCAN_OUT) {
+			res = usb_control_msg(usb, FROM_DEV, IDBG_JTAG_FETCH,
+			    end-bit, 0, buffer, bytes, 1000);
+			if (res < 0) {
+				LOG_ERROR("IDBG_JTAG_FETCH failed: %i", res);
+				exit(1);
+			}
+		}
+		bit = end;
+		buffer += MAX_SCAN_BYTES;
+	}
+	cur_state = ir_scan ? TAP_IRPAUSE : TAP_DRPAUSE;
+	idbg_state_move(end_state);
+}
+
+
+static int idbg_execute_queue(void)
+{
+	jtag_command_t *cmd;
+	enum tap_state state;
+	enum scan_type type;
+	int scan_size;
+	u8 *buffer;
+	int res = ERROR_OK;
+
+	for (cmd = jtag_command_queue; cmd; cmd = cmd->next) {
+		switch (cmd->type) {
+		case JTAG_END_STATE:
+			state = cmd->cmd.end_state->end_state;
+			if (state != -1)
+				end_state = state;
+			break;
+		case JTAG_RESET:
+			if (cmd->cmd.reset->trst)
+				cur_state = TAP_RESET;
+			idbg_reset(cmd->cmd.reset->trst, cmd->cmd.reset->srst);
+			break;
+		case JTAG_RUNTEST:
+			state = cmd->cmd.runtest->end_state;
+			if (state != -1)
+				end_state = state;
+			idbg_runtest(cmd->cmd.runtest->num_cycles);
+			break;
+		case JTAG_STATEMOVE:
+			state = cmd->cmd.statemove->end_state;
+			if (state != -1)
+				end_state = state;
+			idbg_state_move(end_state);
+			break;
+		case JTAG_PATHMOVE:
+			idbg_path_move(cmd->cmd.pathmove);
+			break;
+		case JTAG_SCAN:
+			state = cmd->cmd.scan->end_state;
+			if (state != -1)
+				end_state = state;
+			scan_size = jtag_build_buffer(cmd->cmd.scan, &buffer);
+			type = jtag_scan_type(cmd->cmd.scan);
+			idbg_scan(cmd->cmd.scan->ir_scan, type, buffer,
+			    scan_size);
+			if (jtag_read_buffer(buffer, cmd->cmd.scan) != ERROR_OK)
+				res = ERROR_JTAG_QUEUE_FAILED;
+			if (buffer)
+				free(buffer);
+			break;
+		case JTAG_SLEEP:
+			jtag_sleep(cmd->cmd.sleep->us);
+			break;
+		default:
+			LOG_ERROR("unknown JTAG command %d", cmd->type);
+			exit(1);
+		}
+	}
+	return res;
+}
+
+
+static int idbg_speed(int speed)
+{
+	return ERROR_OK;
+}
+
+
+static int idbg_register_commands(struct command_context_s *cmd_ctx)
+{
+	/* @@@ set vid/pid */
+	return ERROR_OK;
+}
+
+
+static int idbg_init(void)
+{
+	int res;
+
+	usb = open_usb();
+	if (!usb) {
+		LOG_ERROR("no IDBG found");
+		return ERROR_JTAG_INIT_FAILED;
+	}
+	res = usb_control_msg(usb, TO_DEV, IDBG_JTAG_ATTACH, 0, 0,
+	    NULL, 0, 1000);
+	if (res < 0) {
+		LOG_ERROR("IDBG_JTAG_ATTACH failed: %i", res);
+		return ERROR_JTAG_INIT_FAILED;
+	}
+	LOG_INFO("IDBG JTAG interface ready");
+	idbg_reset(0, 0);
+	idbg_write(0, 0, 0);
+	idbg_move(0x7f);
+	return ERROR_OK;
+}
+
+
+static int idbg_quit(void)
+{
+	int res;
+
+	res = usb_control_msg(usb, TO_DEV, IDBG_JTAG_DETACH, 0, 0,
+	    NULL, 0, 1000);
+	if (res < 0) {
+		LOG_ERROR("IDBG_JTAG_DETACH failed: %i", res);
+		return ERROR_JTAG_DEVICE_ERROR;
+	}
+	res = usb_close(usb);
+	if (res < 0) {
+		LOG_ERROR("IDBG_JTAG_ATTACH failed");
+		return ERROR_JTAG_DEVICE_ERROR;
+	}
+	return ERROR_OK;
+}
+
+
+jtag_interface_t idbg_interface = {
+	.name	= "idbg",
+	.execute_queue
+		= idbg_execute_queue,
+	.speed	= idbg_speed,
+	.register_commands
+		= idbg_register_commands,
+	.init	= idbg_init,
+	.quit	= idbg_quit,
+};
Index: openocd/src/jtag/jtag.c
===================================================================
--- openocd/src/jtag/jtag.c	(revision 1279)
+++ openocd/src/jtag/jtag.c	(working copy)
@@ -200,6 +200,10 @@
 	extern jtag_interface_t rlink_interface;
 #endif
 
+#if BUILD_IDBG == 1
+	extern jtag_interface_t idbg_interface;
+#endif
+
 jtag_interface_t *jtag_interfaces[] = {
 #if BUILD_ECOSBOARD == 1
 	&zy1000_interface,
@@ -240,6 +244,9 @@
 #if BUILD_RLINK == 1
 	&rlink_interface,
 #endif
+#if BUILD_IDBG == 1
+	&idbg_interface,
+#endif
 	NULL,
 };
 
Index: openocd/src/jtag/Makefile.am
===================================================================
--- openocd/src/jtag/Makefile.am	(revision 1279)
+++ openocd/src/jtag/Makefile.am	(working copy)
@@ -96,7 +96,13 @@
 RLINKFILES =
 endif
 
+if IDBG
+IDBGFILES = idbg.c
+else
+IDBGFILES =
+endif
+
 libjtag_a_SOURCES = jtag.c $(BITBANGFILES) $(PARPORTFILES) $(DUMMYFILES) $(FT2232FILES) $(AMTJTAGACCELFILES) $(EP93XXFILES) \
-	$(AT91RM9200FILES) $(GW16012FILES) $(BITQFILES) $(PRESTOFILES) $(USBPROGFILES) $(ECOSBOARDFILES) $(JLINKFILES) $(RLINKFILES)
+	$(AT91RM9200FILES) $(GW16012FILES) $(BITQFILES) $(PRESTOFILES) $(USBPROGFILES) $(ECOSBOARDFILES) $(JLINKFILES) $(RLINKFILES) $(IDBGFILES)
 
 noinst_HEADERS = bitbang.h jtag.h
Index: openocd/src/Makefile.am
===================================================================
--- openocd/src/Makefile.am	(revision 1279)
+++ openocd/src/Makefile.am	(working copy)
@@ -50,10 +50,14 @@
 if RLINK
 LIBUSB = -lusb
 else
+if IDBG
+LIBUSB = -lusb
+else
 LIBUSB =
 endif
 endif
 endif
+endif
 
 
 
